<launch>
    <arg name="onboard" default="false"/>
    <group if="$(arg onboard)">	
        <!-- ORB-SLAM3 -->
        <node name="orb_slam3_stereo" pkg="orb_slam3_ros_wrapper" type="orb_slam3_ros_wrapper_stereo" output="screen">
            <!-- for Realsense D435i -->
            <remap from="/camera/left/image_raw"    to="/camera/infra1/image_rect_raw"/>
            <remap from="/camera/right/image_raw"   to="/camera/infra2/image_rect_raw"/>

            <!-- Parameters for original ORB-SLAM3 -->
            <param name="voc_file"      type="string"   value="$(find orb_slam3_ros_wrapper)/config/ORBvoc.bin" />
            <param name="settings_file" type="string"   value="$(find orb_slam3_ros_wrapper)/config/D435i_stereo_inertial_non_rectified.yaml" />
            <param name="do_rectify"    type="bool"     value="false" />
            <param name="do_equalize"   type="bool"     value="false" />

            <!-- Parameters for ROS -->
            <param name="use_viewer"   type="bool"     value="false" />
            <param name="publish_tf_transform"   type="bool"     value="false" />
            <param name="map_frame_id"  type="string"   value="world" />
            <param name="pose_frame_id" type="string"   value="camera_link" />
        </node>
    </group>

    <group unless="$(arg onboard)">
        <!-- republish topic from compressed image -->
        <node name="republish_infra1" type="republish" pkg="image_transport" args="compressed in:=/camera/infra1/image_rect_raw raw out:=/infra1" />
        <node name="republish_infra2" type="republish" pkg="image_transport" args="compressed in:=/camera/infra2/image_rect_raw raw out:=/infra2" />
            
        <node name="orb_slam3_stereo" pkg="orb_slam3_ros_wrapper" type="orb_slam3_ros_wrapper_stereo" output="screen">
            <!-- for Realsense D435i -->
            <remap from="/camera/left/image_raw"    to="/infra1"/>
            <remap from="/camera/right/image_raw"   to="/infra2"/>

            <!-- Parameters for original ORB-SLAM3 -->
            <param name="voc_file"      type="string"   value="$(find orb_slam3_ros_wrapper)/config/ORBvoc.bin" />
            <param name="settings_file" type="string"   value="$(find orb_slam3_ros_wrapper)/config/D435i_stereo_inertial_non_rectified.yaml" />
            <param name="do_rectify"    type="bool"     value="false" />
        
            <!-- Parameters for ROS -->
            <param name="use_viewer"   type="bool"     value="true" />
            <param name="publish_tf_transform"   type="bool"     value="false" />
            <param name="map_frame_id"  type="string"   value="world" />
            <param name="pose_frame_id" type="string"   value="camera_link" />
        </node>
        
        <!-- Visualization - RViz-->
        <node name="rviz" pkg="rviz" type="rviz" args="-d $(find orb_slam3_ros_wrapper)/config/orb_slam3.rviz" output="screen" />

        <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_orb_slam3" output="screen" ns="hector_trajectory_server" >
            <param name="/target_frame_name" value="/world" />
            <param name="/source_frame_name" value="/camera_link" />
            <param name="/trajectory_update_rate" value="20.0" />
            <param name="/trajectory_publish_rate" value="20.0" />
        </node>
    </group>
</launch>
